service-runner.js

```diff
@@ -215,25 +256,71 @@ ServiceRunner.prototype._rollingRestart = function() {
     });
 };
 
+/*
+    This is a workaround for the following bug in node:
+    https://github.com/joyent/node/issues/9409
+
+    Shortly, the problem is that a worker is removed by master
+    if it sends 'disconnect' and then it's checked on 'exit'.
+    However, in some cases (for example an infinite CPU loop in worker,
+    'disconnect' may never happen, and master will crash receiving an 'exit'
+ */
+function fixCloseDisconnectListeners(worker) {
+    // Take the exit listener node's cluster have set.
+    var listeners = worker.process.listeners('exit')[0];
+    var exit = listeners[Object.keys(listeners)[0]];
+
+    // Take the disconnect litener node's cluster have set.
+    listeners = worker.process.listeners('disconnect')[0];
+    var disconnect = listeners[Object.keys(listeners)[0]];
+
+    // Now replace the exit listener to make sure 'disconnect' was called
+    worker.process.removeListener('exit', exit);
+    worker.process.once('exit', function(exitCode, signalCode) {
+        if (worker.state != 'disconnected') {
+            disconnect();
+        }
+        exit(exitCode, signalCode);
+    });
+}
+
 // Fork off one worker at a time, once the previous worker has finished
 // startup.
-ServiceRunner.prototype._startWorkers = function(remainingWorkers, msg) {
+ServiceRunner.prototype._startWorkers = function(remainingWorkers) {
```

Comment:
Maybe it would make sense to set up a worker's `workerHeartbeatTime` right away when forking it so that we are sure the first heartbeat happens in a timely fashion?